src/bp/nlu-core/entities/microsoft-extractor/index.ts

```diff
@@ -0,0 +1,380 @@
+import lru from 'lru-cache'
+import ms from 'ms'
+import sizeof from 'object-sizeof'
+import path from 'path'
+
+import { EntityExtractionResult, SystemEntityExtractor, KeyedItem } from '../../typings'
+import { SystemEntityCacheManager } from '../entity-cache-manager'
+
+type SupportedLangs = 'zh' | 'nl' | 'en' | 'fr' | 'de' | 'it' | 'ja' | 'pt' | 'es'
+const SupportedLangsList = ['zh', 'nl', 'en', 'fr', 'de', 'it', 'ja', 'pt', 'es']
+
+interface MicrosoftParams {
+  lang: SupportedLangs
+  recognizers: any[]
+}
+
+const BATCH_SIZE = 10000
+
+// Further improvements:
+// 1- in _extractBatch, shift results ==> don't walk whole array n times (nlog(n) vs n2)
+
+export class MicrosoftEntityExtractor implements SystemEntityExtractor {
+  private _cache: SystemEntityCacheManager
+
+  constructor(private readonly logger?: NLU.Logger) {
+    this.logger = logger
+    this._cache = new SystemEntityCacheManager(
+      path.join(process.APP_DATA_PATH || '', 'cache', 'microsoft_sys_entities.json'),
+      true,
+      logger
+    )
+  }
+
+  private _langToCulture = (lang: SupportedLangs): string => {
+    switch (lang) {
+      case 'zh':
+        return Recognizers.Culture.Chinese
+      case 'nl':
+        return Recognizers.Culture.Dutch
+      case 'en':
+        return Recognizers.Culture.English
+      case 'fr':
+        return Recognizers.Culture.French
+      case 'de':
+        return Recognizers.Culture.German
+      case 'it':
+        return Recognizers.Culture.Italian
+      case 'ja':
+        return Recognizers.Culture.Japanese
+      case 'pt':
+        return Recognizers.Culture.Portuguese
+      case 'es':
+        return Recognizers.Culture.Spanish
+      default:
+        throw new Error('This language is not supported by Microsoft Recognizer')
+    }
+  }
+
+  private _globalRecognizers = [
+    Recognizers.recognizePhoneNumber,
+    Recognizers.recognizeIpAddress,
+    Recognizers.recognizeMention,
+    Recognizers.recognizeHashtag,
+    Recognizers.recognizeEmail,
+    Recognizers.recognizeURL,
+    Recognizers.recognizeGUID
+  ]
+
+  private _languageDependantRecognizers = [
+    Recognizers.recognizeNumber,
+    Recognizers.recognizeOrdinal,
+    Recognizers.recognizePercentage,
+    Recognizers.recognizeAge,
+    Recognizers.recognizeCurrency,
+    Recognizers.recognizeDimension,
+    Recognizers.recognizeTemperature,
+    Recognizers.recognizeDateTime,
+    Recognizers.recognizeBoolean
+  ]
+
+  public static get entityTypes(): string[] {
+    return []
+  }
+
+  public async configure() {
+    await this._cache.restoreCache()
+  }
+
+  public async extractMultiple(
+    inputs: string[],
+    lang: string,
+    useCache?: boolean
+  ): Promise<EntityExtractionResult[][]> {
+    let options: MicrosoftParams
+
+    if (!SupportedLangsList.includes(lang)) {
+      lang = 'en'
+      options = {
+        lang: 'en' as SupportedLangs,
+        recognizers: [this._globalRecognizers]
+      }
+    }
+
+    options = {
```

Comment:
this will override the option object you define inside `if (!SupportedLangsList.includes(lang)) {`

Reply:
(this should result in a automated test failing)